packages/json-rpc-engine/src/JsonRpcEngine.ts

```diff
@@ -211,15 +217,26 @@ export class JsonRpcEngine extends SafeEventEmitter {
 
     if (Array.isArray(req)) {
       if (callback) {
-        return this.#handleBatch(req, callback);
+        return this.#handleBatch(
+          req,
+          // This assertion is safe because of the runtime checks validating that `req` is an array and `callback` is defined.
+          // There is only one overload signature that satisfies both conditions, and its `callback` type is the one that's being asserted.
+          callback as (
+            error: unknown,
+            responses?: JsonRpcResponse<Json>[],
+          ) => void,
+        );
       }
       return this.#handleBatch(req);
     }
 
     if (callback) {
-      return this.#handle(req as JsonRpcRequest, callback);
+      return this.#handle(
+        req,
+        callback as (error: unknown, response?: JsonRpcResponse<Json>) => void,
```

Comment:
The first and second overload signatures of `handle` are the only cases where `req` is a non-array and `callback` is defined. This is assertion is safe for those overload signatures.